bitcoin/src/blockdata/witness.rs

```diff
@@ -666,6 +702,23 @@ mod test {
         assert_eq!(witness.to_vec(), expected_witness);
     }
 
+    #[test]
+    fn test_push_int() {
+        let mut witness = Witness::new();
+
+        witness.push_int(1); assert_eq!(witness.last().unwrap(), &[0x51]);
+        witness.push_int(0); assert_eq!(witness.last().unwrap(), &[0x00]);
+        witness.push_int(16); assert_eq!(witness.last().unwrap(), &[0x60]);
```

Comment:
nit: it took me a bit of digging to work out what these values are. To make life easier for the next code reader perhaps add some comments, eg  

```suggestion
        witness.push_int(1); assert_eq!(witness.last().unwrap(), &[0x51]); // OP_PUSHNUM_1
        witness.push_int(0); assert_eq!(witness.last().unwrap(), &[0x00]); 
        witness.push_int(16); assert_eq!(witness.last().unwrap(), &[0x60]); // OP_PUSHNUM_16
```

Reply:
0x51 in a witness field is **not** the same as PUSHNUM_1. 0x51 is the number 0x51. PUSHNUM_1 pushes the number 1.

Reply:
Looking at Core CScriptNum I don't see the length byte thing, are these test correct? Said another way, I don't know _why_ we do not add the length byte when encoding integers that are not in range `1..=16` (and -1) but for other integers we do.

These tests all pass, @apoelstra can you confirm this is as expected please?

```rust
        let mut witness = Witness::new();

        // Values that have an OP_PUSHNUM don't use length byte.
        witness.push_int(1); assert_eq!(witness.last().unwrap(), &[0x51]); // OP_PUSHNUM_1
        witness.push_int(0); assert_eq!(witness.last().unwrap(), &[0x00]);
        witness.push_int(16); assert_eq!(witness.last().unwrap(), &[0x60]); // OP_PUSHNUM_16
        witness.push_int(-1); assert_eq!(witness.last().unwrap(), &[0x4f]); // OP_PUSHNUM_NEG1

        // For other values the length byte is the length of the encoded data as well as the sign byte.
        //
        // Decimal 81 is 0x51 in hex
        witness.push_int(81); assert_eq!(witness.last().unwrap(), &[0x01, 0x51]); 
        
        // Top bit is set, and value is positive, so append 0x00.
        witness.push_int(128); assert_eq!(witness.last().unwrap(), &[0x02, 0x80, 0x00]); 
        // Top bit is set, and value is negative, so append 0x80.
        witness.push_int(-128); assert_eq!(witness.last().unwrap(), &[0x02, 0x80, 0x80]);

        witness.push_int(17); assert_eq!(witness.last().unwrap(), &[0x01, 0x11]); // Decimal 17 is 0x11 or 0b0001_0001
        witness.push_int(-17); assert_eq!(witness.last().unwrap(), &[0x01, 0x91]); // Turn on top bit, 0x91 or 0b1001_0001

        witness.push_int(u32::MAX.into()); assert_eq!(witness.last().unwrap(), &[0x05, 0xff, 0xff, 0xff, 0xff, 0x00]);
        witness.push_int((u32::MAX as i64) * -1); assert_eq!(witness.last().unwrap(), &[0x05, 0xff, 0xff, 0xff, 0xff, 0x80]);

        // serialized varint in little-endian signed-magnitude format preceded by length byte
        witness.push_int(10000); assert_eq!(witness.last().unwrap(), &[0x02, 0x10, 0x27]);
        witness.push_int(255); assert_eq!(witness.last().unwrap(), &[0x02, 0xff, 0x00]);
        witness.push_int(36507); assert_eq!(witness.last().unwrap(), &[0x03, 0x9b, 0x8e, 0x00]);

        witness.push_int(-2); assert_eq!(witness.last().unwrap(), &[0x01, 0x82]);
        witness.push_int(-3456748); assert_eq!(witness.last().unwrap(), &[0x03, 0xec, 0xbe, 0xb4]);
        witness.push_int(-10000000); assert_eq!(witness.last().unwrap(), &[0x04, 0x80, 0x96, 0x98, 0x80]);
```

Reply:
These pretty-much all look wrong, or at least confused.

There should be no length bytes inside witness elements. The witness elements themselves are length prefixed.

And OP_PUSHNUM_1 etc have *nothing* to do with witness-encoded numbers, which is what I would expect this function to be producing. 0x51 encodes the value 0x51. Yes, this is numerically equal to the PUSHNUM_1 opcode but there is no possible way to interpret a witness element as an opcode, and the impression a reader gets is that you were trying to push the value 1, not 0x51.

Reply:
Just giving my take, I used [Builder's test](https://github.com/rust-bitcoin/rust-bitcoin/blob/6a2fd96ff6087f1f9fc979d8639dc1ff6d2ea693/bitcoin/src/blockdata/script/tests.rs#L15C1-L48C2) to derive the integers. Also, I myself don't really know whether it should have a length component (personally I think it shouldn't), but I strictly tried to adhere to the builder's push_int method so I don't make a mistake.

Reply:
@startup-dreamer I apologize but I don't have time to guide somebody through writing this PR if they don't already know the details. I think this is not a good task for a new contributor. At some point in the future I will implement this method.

>Do you mean the tests are wrong in that they should not pass or that they are confused as in the test cases are confused (and/or the code comments)?

I didn't read them all, but I glanced at 5 or 6 and every one was at least one of those.

Reply:
No worries, still, thanks for your time. I will try to learn more in depth and implement smaller things first. Thanks for the review.

Reply:
Don't be put off. There are loads of things to do that do not require deep knowledge of Bitcoin and it is not always obvious what issues do require it and what issues don't. Try to push PRs that you know are correct because you really understand the changes, in otherwords try to understand _why_ the code does what it does and only make changes when you fully understand the "why".